multiwallet.go

```diff
@@ -0,0 +1,518 @@
+	if strings.HasPrefix(walletName, "wallet-") {
+		return false, errors.E(ErrReservedWalletName)
+	}
+
+	err := mw.db.One("Name", walletName, &Wallet{})
+	if err == nil {
+		return true, nil
+	} else if err != storm.ErrNotFound {
+		return false, err
+	}
+
+	return false, nil
+}
+
+func (mw *MultiWallet) OpenWallets(pubPass []byte) error {
+	if mw.syncData.activeSyncData != nil {
+		return errors.New(ErrSyncAlreadyInProgress)
+	}
+
+	for _, wallet := range mw.wallets {
+		err := wallet.OpenWallet(pubPass)
+		if err != nil {
+			return err
+		}
+
+		go mw.listenForTransactions(wallet.ID)
+	}
+
+	return nil
+}
+
+func (mw *MultiWallet) OpenWallet(walletID int, pubPass []byte) error {
+	if mw.syncData.activeSyncData != nil {
+		return errors.New(ErrSyncAlreadyInProgress)
+	}
+
+	wallet := mw.WalletWithID(walletID)
+	if wallet == nil {
+		return errors.New(ErrNotExist)
+	}
+
+	err := wallet.OpenWallet(pubPass)
+	if err != nil {
+		return err
+	}
+
+	go mw.listenForTransactions(wallet.ID)
+	return nil
+}
+
+func (mw *MultiWallet) UnlockWallet(walletID int, privPass []byte) error {
+	wallet := mw.WalletWithID(walletID)
+	if wallet == nil {
+		return errors.New(ErrNotExist)
+	}
+
+	return wallet.UnlockWallet(privPass)
+}
+
+func (mw *MultiWallet) ChangeStartupPassphrase(oldStartupPass, newStartupPass []byte) error {
+	defer func() {
+		for i := range oldStartupPass {
+			oldStartupPass[i] = 0
+		}
+
+		for i := range newStartupPass {
+			newStartupPass[i] = 0
+		}
+	}()
+
+	if len(oldStartupPass) == 0 {
+		oldStartupPass = []byte(w.InsecurePubPassphrase)
+	}
+	if len(newStartupPass) == 0 {
+		newStartupPass = []byte(w.InsecurePubPassphrase)
+	}
+
+	successfullyChangedWalletIDs := make([]int, 0)
+	var err error
+	for walletID, wallet := range mw.wallets {
+		ctx, _ := mw.contextWithShutdownCancel()
+		if err = wallet.internal.ChangePublicPassphrase(ctx, oldStartupPass, newStartupPass); err != nil {
+			log.Errorf("[%d] Error changing public passphrase: %v", walletID, err)
+			break
+		}
+		successfullyChangedWalletIDs = append(successfullyChangedWalletIDs, walletID)
+	}
+
+	if err != nil {
+		// Rollback changes
+		for walletID := range successfullyChangedWalletIDs {
+			ctx, _ := mw.contextWithShutdownCancel()
+			mw.wallets[walletID].internal.ChangePublicPassphrase(ctx, newStartupPass, oldStartupPass)
+		}
+	}
+
+	return translateError(err)
+}
+
+func (mw *MultiWallet) ChangePrivatePassphrase(walletID int, oldPrivatePassphrase, newPrivatePassphrase []byte, privatePassphraseType int32) error {
+	if privatePassphraseType != PassphraseTypePin && privatePassphraseType != PassphraseTypePass {
+		return errors.New(ErrInvalid)
+	}
+
+	wallet := mw.WalletWithID(walletID)
+	if wallet == nil {
+		return errors.New(ErrInvalid)
+	}
+
+	err := wallet.changePrivatePassphrase(oldPrivatePassphrase, newPrivatePassphrase)
+	if err != nil {
+		return translateError(err)
+	}
+
+	wallet.PrivatePassphraseType = privatePassphraseType
```

Comment:
`PassphraseType` is used in the frontend to determine what input type should be shown to the user. Public passphrase type is not saved with the other wallet properties because it is the same for all wallets so instead, it is saved as a config.